sky/skylet/providers/aws/config_v2.py

```diff
@@ -0,0 +1,1234 @@
+    )  # todo: err msg
+    assert profile is not None, "Failed to create instance profile"
+
+    if not profile.roles:
+        role_name = cwh.resolve_iam_role_name(config["provider"], DEFAULT_SKYPILOT_IAM_ROLE)
+        role = _get_role(role_name, config)
+        if role is None:
+            cli_logger.verbose(
+                "Creating new IAM role {} for use as the default instance role.",
+                cf.bold(role_name),
+            )
+            iam = _resource("iam", config)
+            policy_doc = {
+                "Statement": [
+                    {
+                        "Effect": "Allow",
+                        "Principal": {"Service": "ec2.amazonaws.com"},
+                        "Action": "sts:AssumeRole",
+                    },
+                ]
+            }
+            attach_policy_arns = cwh.resolve_policy_arns(
+                config["provider"],
+                iam,
+                [
+                    "arn:aws:iam::aws:policy/AmazonEC2FullAccess",
+                    "arn:aws:iam::aws:policy/AmazonS3FullAccess",
+                ],
+            )
+            iam.create_role(
+                RoleName=role_name, AssumeRolePolicyDocument=json.dumps(policy_doc)
+            )
+            
+            role = _get_role(role_name, config)
+            cli_logger.doassert(
+                role is not None, "Failed to create role."
+            )  # todo: err msg
+
+            assert role is not None, "Failed to create role"
+
+            for policy_arn in attach_policy_arns:
+                role.attach_policy(PolicyArn=policy_arn)
+            
+            # SkyPilot: "PassRole" is required by the head node to pass the role to
+            # the workers, so we can access S3 buckets on the workers. "Resource"
+            # is to limit the role to only able to pass itself to the workers.
+            iam_self_read_write_policy_doc = {
+                "Statement": [
+                    {
+                        "Effect": "Allow",
+                        "Action": [
+                            "iam:GetRole",
+                            "iam:PassRole",
+                        ],
+                        "Resource": role.arn,
+                    },
+                    {
+                        "Effect": "Allow",
+                        "Action": "iam:GetInstanceProfile",
+                        "Resource": profile.arn,
+                    }
+                ]
+            }
+            role.Policy("IAMSelfReadWritePolicy").put(
+                PolicyDocument=json.dumps(iam_self_read_write_policy_doc)
+            )
+
+        profile.add_role(RoleName=role.name)
+        time.sleep(15)  # wait for propagation
+    # Add IAM role to "head_node" field so that it is applied only to
+    # the head node -- not to workers with the same node type as the head.
+    # config["head_node"]["IamInstanceProfile"] = {"Arn": profile.arn}
```

Comment:
The effect should be similar according to the code below. TL;DR ray looks for the instance profile in the `node_config` first and then goes to the `head_node`, which treats them equivalently. We use the `node_config` since it is general for both head and workers when we don't need to distinguish head or worker.

https://github.com/skypilot-org/skypilot/blob/fa02be38ac8d885bee2c76817b959abf40366d49/sky/skylet/providers/aws/config.py#L191-L203